### ISSUE TITLE
Interpolated string handler: inconsistent use of AppendFormatted when appending constants

### ISSUE THREAD

**Issue by mgravell** [dotnet/roslyn#76391]
Context: custom interpolated string handler, with the intent to perform bespoke operations (think "SQL auto parameterization")

If (and only if) all formatter tokens are constant strings, the string is reduced to a constant expression *instead of* using the expected `AppendFormatted` calls.

**Version Used**: 

.NET 9.0.100 (current)

**Steps to Reproduce**:

The first `Utils.Write` here is the unexpected outcome; contrast this to a: the similar usage including a constant integer, and b: the similar usage using a variable string, or a variable string and variable integer (which all use `AppendFormatted`)

``` c#
using System;
using System.Runtime.CompilerServices;

// constants
const int c_id = 42;
const string c_name = "abc";

// variables
int v_id = c_id;
string v_name = c_name;

// constants, string only
Utils.Write($"c_name: {c_name}"); // <== STRING!!
// constants, string and integer
Utils.Write($"c_id: {c_id}, c_name: {c_name}"); // (interpolated)

// variables, string only
Utils.Write($"v_name: {v_name}"); // (interpolated)
// variables, string and integer
Utils.Write($"v_id: {v_id}, v_name: {v_name}"); // (interpolated)

public static class Utils
{
    public static void Write(ref CustomInterpolatedStringHandler value)
        => Console.WriteLine("Interpolated");
    public static void Write(string value)
        => Console.WriteLine("String");
}
[InterpolatedStringHandler]
public ref struct CustomInterpolatedStringHandler(int x, int y)
{
    public void AppendLiteral(string value) { }
    public void AppendFormatted<T>(T value) { }
}
```

**Actual Behavior**:

The first call uses `Write(string)` with the value concatenated at compile, [as shown on sharplab.io](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABATARgLABQGADAAQY4B0ASgK4B2ALgJYC2MlAwhKwA7MAbGFADKwgG7MwMAM4BuQoQD0S0pHozGAQyYzC6zaWZM1AfWYATUgF5SKLAoIHG5HGTCn6W9jdIAiLWAwP0dlVXEtKGZAoT0CYxdxcytbD0tHCjIkrx9Uz28YUIIVNQgNbV0AGlcyMoEAT0IAVRYBGUoAdSjGGAAKABI/DxyYEFIAb2GCgF8/AEo5UhKAHmtbEQAVagBJADkAcQBCQ7DS8p1GGWrM0h0rBJgAc2Fm1vau5h6BoeSxyeTptUpuw/sCYLMFktVL0HlBeBABFoehY5opiuFItFgLFrm5SHVGgQWoJ3t0+oNsgU/pT2BDFiUYUxhPDEcjUejSBEojFZLiyHcjEznlBXiTOmTvklLNSAdUaaMJvK6VDSIyenCEUiYCi0RgAMyuJDkLCkYltQjjQika3kA0UI0YFCkD5fWAAM1InFomh42yZGtZ2pEjCi9EeAAk7kIoJytAJaDB2TbkzYAHyuACcvT8fvVLK1FnmjmT+sN5CdLr6Nwi8cTVpT1us6YoWb8wdDjyLhGmhAA2rnmZrke3jBGo8IALqEUvu0iaKC0MAuL0+1gDgMFkdhyP0CzRtWkBDVBKkersy0EEt2p0AQV4vBgu4AMp9hHHetW4wm5hNSD3LzapaOqQd4PruABi0CsEiyLLBsqa9Bssa1j+4x/t2hBAA)

**Expected Behavior**:

Reduction occurs *later*; there's a question here as to whether it should still fold and prefer `string` in the case then the handler is a `DefaultInterpolatedStringHandler` and a similar `string` API exists.

**Impact**:

Custom in `AppendFormatted` scenarios are not reliable if the token turns out to be constant. In particular, it means that it is hard to see whether this resolves to `DoThing(string)` or `DoThing(ref CustomInterpolatedStringHandler)`, which may have significant impact if the intent of `CustomInterpolatedStringHandler` is complex

``` c#
obj.DoThing($"""
    some complex
    code with {Formatted} and
    non formatted {Could.Be.Far.Away.Regions} that
    may be miles away
    """);
```

(are `Formatted` and `Regions` constant expressions? we don't know, and neither does the reader)

**Alternative Thought**:

Instead of "fixing" this, maybe there should be a compiler warning if this could be a risk; perhaps all of:

1. an interpolated string is encountered
2. the interpolated string is being reduced to a constant expression by the compiler
3. the receiving argument is *not* `string` or `DefaultInterpolatedStringHandler` 

However, I guess that "3" depends on overload resolution having already occurred; maybe it should be

3. there is a competing usable overload that *takes* an interpolated string handler other than `DefaultInterpolatedStringHandler`

**Comment by jaredpar** [comment]
> The first call uses Write(string) with the value concatenated at compile ...

This was an [explicit design decision](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-10.0/improved-interpolated-strings.md#better-conversion-from-expression-adjustments) for custom interpolated strings. Essentially, the conversion to `string` is better because it's constant. The expectation being that if you overloaded on the two that the custom interpolation handler is just building a `string` hence better to call that directly vs. indirectly building up the string. 

> Custom in AppendFormatted scenarios are not reliable if the token turns out to be constant. 

This is only true though when there is an overload that takes a `string` and the `CustomInterpolatedStringHandler`. If there is only an overload of `CustomInterpolatedStringHandler ` then `AppendFormat` will be called in both cases.

> Instead of "fixing" this, maybe there should be a compiler warning if this could be a risk; 

I worry about a warning here though because this was done based on real APIs that we saw. Hence potentially end up adding a breaking change to other consumers who are happy with this behavior. 

@333fred

**Comment by mgravell** [comment]
@jaredpar thanks; it looks like this can also be influenced by poking it with a `[OverloadResolutionPriority]`-shaped pointy stick, i.e. after adding `[OverloadResolutionPriority(42)]` on the custom handler overload: it gets preferred. If this is expected to be a reliable workaround, it is probably sufficient.

**Comment by 333fred** [comment]
> If this is expected to be a reliable workaround, it is probably sufficient.

Yes, that is a reliable workaround. We expect that, in 99% of cases, API authors who expose both a `string` overload and a custom handler overload would rather have the performance benefits of just using a `string`, when that `string` can calculated entirely at compile time with no runtime allocations. If that's not what you want for your scenario, either not exposing the `string` overload or using `OverloadResolutionPriority` is an appropriate solution.
